[CLEAN-UP] Removes quiz iframe structure

TiBiBa opened the pull request:
First, provide a one line summary of your changes in the **Title of the PR**

**Now fill out the remainder of this template by replacing all _italic_ content**

**Description**

_Changes in detail_

For example: "**Adds** translations of levels 1 to 12 to Polish" (note the use of present tense!)

**Fixes _issue or discussion number_**

Always link the number of the issue or of the discussion that your pr concerns.

**How to test**

* If this is a UI change: _describe how to run your code to see it in action. See this https://github.com/Felienne/hedy/pull/880#issue-1016304308 for an example_
* If this is a language change: _add a few tests showing the difference_

**Checklist**
Done? Check if you have it all in place using this list:*
  
- [ ] Describes changes in the format above (present tense)
- [ ] Links to an existing issue or discussion 
- [ ] Has a "How to test" section

If you're unsure about any of these, don't hesitate to ask. We're here to help!

TiBiBa commented:
Some bugs we are still struggling with:

- Dynamically load answers into either the ace editors or plain text
- Syntax highlighting on the question is broken
- Result page is not working (yet)
- We currently support only one attempt, do we want to change this?
   - If so, we want to track this on the back-end to prevent altering by the user
   - I personally prefer the "one attempt" approach -> its multiple-choice

Felienne commented:
> We currently support only one attempt, do we want to change this?
> * If so, we want to track this on the back-end to prevent altering by the user
> * I personally prefer the "one attempt" approach -> its multiple-choice

One try for me is fine. What happens when you try the quiz for a 2nd time? Do you see your results page?

Felienne commented:
> The database structure was so complex I'm still not sure if it is still working as it was. It appears the original structure was to create a `None` placeholder in the database before an answer is given. Then the actual answer is appended to this value inside a list. We preserve this list structure but remove the redundant `None` storing. This does have the disadvantage that I'm not sure if this still works as expected with the stats.
> 
> @boryanagoncharenko maybe you can help me figure this out regarding the quiz stats?

If I recall correctly, @rix0rrr also helped with the quiz/database implementation, can you confirm that it works like above with the None as first value?

boryanagoncharenko commented:
The quiz stats do not rely on the `None` placeholder. They are quite basic atm - we only track whether the user has started a quiz and their score (if they completed it). The score is taken from the session, not from the DB.

So your change should have no impact on the stats. I also checked your branch and the quiz stats are working as expected.

TiBiBa commented:
> > We currently support only one attempt, do we want to change this?
> > 
> > * If so, we want to track this on the back-end to prevent altering by the user
> > * I personally prefer the "one attempt" approach -> its multiple-choice
> 
> One try for me is fine. What happens when you try the quiz for a 2nd time? Do you see your results page?

When re-loading the page the quiz is automatically re-started. If I remember correctly we currently keep track of the question the user was at when leaving the quiz, continuing there. We can take a look at implementation this again by checking in the database if there is some incomplete quiz entry for the current level.

TiBiBa commented:
> The quiz stats do not rely on the `None` placeholder. They are quite basic atm - we only track whether the user has started a quiz and their score (if they completed it). The score is taken from the session, not from the DB.
> 
> So your change should have no impact on the stats. I also checked your branch and the quiz stats are working as expected.

Awesome, thanks for checking!

rix0rrr commented:
> I guess, given the answer above and the fact that we now only support 1 try, we might also want to rewrite the list to a singular value (can be in a separate PR of course!)

We could, but I'd say why bother.

* If you write scripts to crunch the data you would need to account for both data types (and importantly: **remember** that there are two data types, which will not be obvious if you just look at a recent sampling of the data).
* If you change your mind in the future and want to allow multiple tries again you'd need to change it again

Felienne commented:
> We could, but I'd say why bother.
> 
> * If you write scripts to crunch the data you would need to account for both data types (and importantly: **remember** that there are two data types, which will not be obvious if you just look at a recent sampling of the data).
> * If you change your mind in the future and want to allow multiple tries again you'd need to change it again

Fair enough, let's leave it as it is then!

Felienne reviewed (changes requested):
If possible, I'd like to have 2 tries per question.

TiBiBa commented:
> If possible, I'd like to have 2 tries per question.

Should be possible! Will require quite some re-writing, but I agree that this is a better implementation from a user perspective.

TiBiBa commented:
> Nice, this seems to make the quiz a lot smoother!
> 
> But now I have tested, I understand you mean 1 try _of a question_ and not of a quiz! I think there is a lot of value in giving kids another try for a question when they make a mistake, that is a good learning opportunity. But 1 extra try (so 2 in total) is enough I think, 3 is a but much. Would it be complex to keep that part of the functioning the same?
> 
> Also, there is a tiny issue with the visualization of code in quiz test in the answer page:
> 
> ![image](https://user-images.githubusercontent.com/1003685/170497091-96bcd797-2e5c-46ce-b656-d6ab569a8fe2.png)

This should be easy to fix, will take a look!

TiBiBa commented:
This should work with two attempts!

Felienne reviewed (changes requested):
Once we solved the weird text selection issue this should be good to go!

Felienne commented:
Ehm, the Ace changes seem to have... removed all spaces when code blocks are in answers :) 

![image](https://user-images.githubusercontent.com/1003685/171172918-dbfc7d95-aa99-4ec9-b426-0b375ede5020.png)

TiBiBa commented:
> Ehm, the Ace changes seem to have... removed all spaces when code blocks are in answers :)
> 
> ![image](https://user-images.githubusercontent.com/1003685/171172918-dbfc7d95-aa99-4ec9-b426-0b375ede5020.png)

Whoops...

TiBiBa commented:
> Ehm, the Ace changes seem to have... removed all spaces when code blocks are in answers :)
> 
> ![image](https://user-images.githubusercontent.com/1003685/171172918-dbfc7d95-aa99-4ec9-b426-0b375ede5020.png)

Fixed!

Felienne reviewed (approved):
Works now with spaces! 

![image](https://user-images.githubusercontent.com/1003685/171174430-7b1e6810-4dc8-4724-be7a-c19408d8f1d5.png)
